src/inmanta/agent/agent_new.py

```diff
@@ -198,48 +198,44 @@ async def start_working(self) -> None:
         await self.scheduler.start()
         self._enable_time_triggers()
 
-    async def stop_working(self) -> None:
+    async def stop_working(self, timeout: float = 0.0) -> None:
         """Stop working, connection lost"""
         if not self.working:
             return
         self.working = False
         self._disable_time_triggers()
-        await self.executor_manager.stop()
         await self.scheduler.stop()
+        await self.executor_manager.stop()
+        await self.executor_manager.join([], timeout=timeout)
+        await self.scheduler.join()
 
     @protocol.handle(methods_v2.update_agent_map)
     async def update_agent_map(self, agent_map: dict[str, str]) -> None:
         # Not used here
         pass
 
-    async def unpause(self, name: str) -> Apireturn:
-        if name != AGENT_SCHEDULER_ID:
-            return 404, "No such agent"
-
-        await self.start_working()
-        return 200
-
-    async def pause(self, name: str) -> Apireturn:
-        if name != AGENT_SCHEDULER_ID:
-            return 404, "No such agent"
-
-        await self.stop_working()
-        return 200
-
     @protocol.handle(methods.set_state)
     async def set_state(self, agent: str, enabled: bool) -> Apireturn:
-        if enabled:
-            return await self.unpause(agent)
+        if agent == AGENT_SCHEDULER_ID:
+            should_be_running = await self.scheduler.should_be_running(agent)
+            if should_be_running:
+                await self.start_working()
+            else:
+                await self.stop_working(timeout=const.EXECUTOR_GRACE_HARD)
         else:
-            return await self.pause(agent)
+            try:
+                await self.scheduler.refresh_agent_state_from_db(name=agent)
+            except LookupError:
+                return 404, f"No such agent: {agent}"
+
+        return 200, f"{agent} has been {'started' if enabled else 'stopped'}"
```

Comment:
The msg `f"{agent} has been {'started' if enabled else 'stopped'}"` relies on `enabled` but `refresh_agent_state_from_db` doesn't. We might get inconsistent behaviour no ?